core/templates/filters/translate.pipe.spec.ts

```diff
@@ -0,0 +1,143 @@
+  from '@angular/core/testing';
+import { EventEmitter, ChangeDetectorRef, Component } from '@angular/core';
+
+import { TranslateService } from 'services/translate.service';
+import { TranslatePipe } from './translate.pipe';
+import { UtilsService } from 'services/utils.service';
+
+
+class MockChangeDetectorRef {
+  markForCheck() {
+    return true;
+  }
+}
+
+class MockTranslateService {
+  langChangeEventEmitter = new EventEmitter();
+  get onLangChange() {
+    return this.langChangeEventEmitter;
+  }
+  translations = {
+    I18n_t_1: 'Hello',
+    I18n_t_2: 'Hello <[val]>',
+    I18n_rogue_1: '<script>alert(\'Oppia\');</script>Hello',
+    I18n_rogue_2: '<oppia-img>Me</oppia-img>Hola'
+  };
+
+  getInterpolatedString(key: string,
+      interpolateParams?: Object) {
+    const str = this.translations[key];
+    if (!str) {
+      return key;
+    }
+    return str;
+  }
+}
+
+describe('TranslatePipe', () => {
+  let pipe: TranslatePipe;
+  let translateService: TranslateService;
+  let changeDecRef: ChangeDetectorRef;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [
+        TranslatePipe
+      ],
+      providers: [
+        { provide: TranslateService, useClass: MockTranslateService },
+        { provide: ChangeDetectorRef, useClass: MockChangeDetectorRef}
+      ]
+    }).compileComponents();
+    translateService = TestBed.get(TranslateService);
+    changeDecRef = TestBed.get(ChangeDetectorRef);
+    pipe = new TranslatePipe(
+      translateService, changeDecRef, new UtilsService());
+  }));
+
+  it('should translate', () => {
+    expect(pipe.transform('I18n_t_1')).toBe('Hello');
+    expect(pipe.transform('I18n_t_2', {val: 'World'})).toBe('Hello <[val]>');
+    expect(pipe.transform('I18n_t_2', {val: 'World'})).toBe('Hello <[val]>');
+    expect(pipe.transform('I18n_t_3')).toBe('I18n_t_3');
+    expect(pipe.transform('')).toBe('');
+    translateService.onLangChange.emit({lang: 'en'});
+  });
+
+  // The sole purpose of this test is to cover ngOnDestroy.
+  it('should destroy subscriptions', () => {
+    pipe.ngOnDestroy();
+    expect(pipe).not.toBeDefined;
+    // Reintializing the pipe because jasmine tries to destroy the pipe.
+    // But since ngOnDestroy method has been called and the pipe is destroyed.
+    // And if the pipe is not reinitialized karma will raise an error saying
+    // "cannot call ngOnDestroy of undefined".
+    pipe = new TranslatePipe(
+      translateService, changeDecRef, new UtilsService());
+  });
+});
+
+// Adding the following two lines for reference.
+// I18n_rogue_1: '<script>alert(\'Oppia\');</script>Hello',
+// I18n_rogue_2: '<oppia-img>Me</oppia-img>Hola'
+@Component({
+  // eslint-disable-next-line no-multi-str, angular/no-inline-template
+  template: '<h1 [innerHTML] = "\'I18n_rogue_1\' | translate"></h1>\
+              <h2 [innerHTML] = "\'I18n_rogue_2\' | translate"></h2>'
+})
+class MockComponent { }
+
+describe('Angular', () => {
```

Comment:
@seanlip this should suffice, right?

Reply:
Here, you are testing rogue i18n (which is great) but you should also have a test for rogue params (like bad values for val) as part of the test above. If it doesn't strip them then add a comment explaining this and outline where the sanitization takes place, so that it's easy for devs to understand the expected behaviour.

Reply:
The sanitization happens for the value that is returned from the pipe. So having them in i18n or params wouldn't make a difference. However, I have added extra tests for rogue params too.

Reply:
The extra tests here are good, but I meant adding the rogue param tests [here](https://github.com/oppia/oppia/pull/9398/files#diff-3c2ff5dc31729ccca47e22d347459d3aR41) too, just to clarify where the sanitization is done (or not done).

Reply:
Added extra tests here!